test/integration/scoped_rds_integration_test.cc

```diff
@@ -732,5 +732,50 @@ on_demand: true
   }
 }
 
+TEST_P(ScopedRdsIntegrationTest, OnDemandUpdateAfterActiveStreamDestroyed) {
+  config_helper_.addFilter(R"EOF(
+    name: envoy.filters.http.on_demand
+    )EOF");
+  const std::string scope_route1 = R"EOF(
+name: foo_scope1
+route_configuration_name: foo_route1
+on_demand: true
+key:
+  fragments:
+    - string_key: foo
+)EOF";
+  on_server_init_function_ = [this, &scope_route1]() {
+    createScopedRdsStream();
+    sendSrdsResponse({scope_route1}, {scope_route1}, {}, "1");
+  };
+  initialize();
+  registerTestServerPorts({"http"});
+  fake_upstreams_[0]->set_allow_unexpected_disconnects(true);
+
+  const std::string route_config_tmpl = R"EOF(
+      name: {}
+      virtual_hosts:
+      - name: integration
+        domains: ["*"]
+        routes:
+        - match: {{ prefix: "/" }}
+          route: {{ cluster: {} }}
+)EOF";
+  codec_client_ = makeHttpConnection(makeClientConnection((lookupPort("http"))));
+  // Request that match lazily loaded scope will trigger on demand loading.
+  auto response = codec_client_->makeHeaderOnlyRequest(
+      Http::TestRequestHeaderMapImpl{{":method", "GET"},
+                                     {":path", "/meh"},
+                                     {":authority", "host"},
+                                     {":scheme", "http"},
+                                     {"Addr", "x-foo-key=foo"}});
+  // Close the connection and destroy the active stream.
+  cleanupUpstreamAndDownstream();
+  // Push rds update and there is noexception thrown.
```

Comment:
why would there be an exception? 

nit: "no exception thrown".

Reply:
Because the active stream is killed. When the updated callback is executed, if we don't try to lock the callback shared ptr, there will be segment fault.

Reply:
There wont be an exception if it's an segfault. was hoping you could clarify in the comment, I think this comment slipped off the table as we are unfamiliar with this UI. 
Could you shoot a quick PR to fix it?

Reply:
Aceesing a destroyed object leads to undefined behavior. It is likely to cause segment fault but not necessary.